### DISCUSSION TITLE
Evaluating and Predicting from a trained SpanCat Model Pipeline

### DISCUSSION THREAD

**Discussion by ghost** [explosion/spaCy#9121] [Help: Coding & Implementations] [answered]
I trained a spacy SpanCat model pipeline using commands given in [spaCy's Offical Documentation](https://spacy.io/api/spancategorizer) but when when I am trying to make predictions it is giving me 2 arrays, one with spans and the other with (i guess) probability which is quite low in 10^-6. 

I am initializing SpanCat using the following commands- 

```
nlp = spacy.load('en_core_web_lg')
nlp.disable_pipes(*[pipe for pipe in nlp.pipe_names if pipe != 'spancat'])
spancat = nlp.add_pipe('spancat')
spancat.add_label(custom_label)
spancat.initialize(get_examples = train_examples, nlp = nlp, labels = spancat.label_data)
```

I am updating SpanCat using the following command- 

`spancat.update(train_examples, sgd = spancat.create_optimizer())`

Also when I try to calculate Precision, Recall, and F-Score using `spancat.score(examples)` I am getting this

`{'spans_sc_p': None, 'spans_sc_r': None, 'spans_sc_f': None, 'spans_sc_per_type': None}`

Any help is much appreciated.

Thanks in advance

**Comment by svlandeg** [comment, marked as answer]
What could help is to try and make this work for just one or two examples, and to see whether you can get the `spancat` model to overfit on them.

What is missing from your code snippet is how you've defined `train_examples` and `examples` - these are crucial parts to getting the training and scoring right. During training, you typically want to make sure that any `Example` object contains the correct `reference` document and a blank `predicted` document. At scoring time, you want to make sure that the `predicted` doc actually does contain the predictions by running the newly trained `nlp` document on the text.

Here's a working code snippet that may help you get on your way:
```
    nlp = spacy.load('en_core_web_lg')
    nlp.disable_pipes(*[pipe for pipe in nlp.pipe_names if pipe != 'spancat'])
    spancat = nlp.add_pipe("spancat")
    spancat.add_label("PERSON")
    text = "Who is Shaka Khan?"
    
    # set up toy gold document and training example
    gold_doc = nlp.make_doc(text)
    gold_span = gold_doc[2:4]
    gold_span.label_ = "PERSON"
    gold_doc.spans["sc"] = [gold_span]
    train_examples = [Example(nlp.make_doc(text), gold_doc)]
    
    optimizer = spancat.initialize(get_examples=lambda: train_examples)
    for i in range(50):
        losses = {}
        nlp.update(train_examples, sgd=optimizer, losses=losses)

    # set up toy test example
    pred_doc = nlp(text)
    test_examples = [Example(pred_doc, gold_doc)]
    
    # print the final results - should be 1.0 for all measures.
    print(spancat.score(test_examples))
```

**Reply by ghost** [reply]
Thanks a lot. My issue was actually with how I defined `train_examples` and `examples`. I provided spans in both the `predicted `and `reference` . Thanks a lot for you help and time. :)
